src/Tools/dotnet-monitor/Egress/AzureBlob/AzureBlobEgressProvider.cs

```diff
@@ -143,6 +153,53 @@ public AzureBlobEgressProvider(ILogger<AzureBlobEgressProvider> logger)
             }
         }
 
+        public async Task SetBlobClientMetadata(BlobBaseClient blobClient, EgressArtifactSettings artifactSettings, CancellationToken token)
+        {
+            Dictionary<string, string> mergedMetadata = new Dictionary<string, string>(artifactSettings.Metadata);
+
+            foreach (var metadataPair in artifactSettings.CustomMetadata)
+            {
+                if (!mergedMetadata.ContainsKey(metadataPair.Key))
+                {
+                    mergedMetadata[metadataPair.Key] = metadataPair.Value;
+                }
+                else
+                {
+                    Logger.DuplicateKeyInMetadata(metadataPair.Key);
+                }
+            }
+
+            try
+            {
+                // Write blob metadata
+                await blobClient.SetMetadataAsync(mergedMetadata, cancellationToken: token);
+            }
+            catch (Exception ex) when (ex is InvalidOperationException || ex is RequestFailedException)
+            {
+                Logger.InvalidMetadata(ex);
+                await blobClient.SetMetadataAsync(artifactSettings.Metadata, cancellationToken: token);
+            }
+        }
+
+        public async Task AddConfiguredMetadataAsync(AzureBlobEgressProviderOptions options, EgressArtifactSettings artifactSettings, IEndpointInfo endpointInfo, CancellationToken token)
+        {
+            DiagnosticsClient client = new DiagnosticsClient(endpointInfo.Endpoint);
+
+            Dictionary<string, string> envBlock = await client.GetProcessEnvironmentAsync(token);
```

Comment:
Rather than having this specific egress provider do this, could the egress service do this and fill in the CustomMetadata? This would eliminate the need to pass the IEndpointInfo into the egress provider, which is had no knowledge of previously.

Reply:
So, I can do that, although I'm not sure that's the correct level for it - we don't currently have any notion of which options type we're using at that level, and then once we get the options type we'd only add custom metadata if the type is Azure (since this doesn't apply to FileSystem). The way it's currently written, `EgressService` is kept generic and metadata is added regardless of the provider, but that would no longer be the case (unless we added a dummy field for FileSystem for custom metadata that is always empty, but that feels a bit off).

Reply:
I was thinking more in terms of the egress provider would have a method/property on it that says "I'm interested in the environment block of the target process", then the egress service would populate the environment block on the EgressArtifactsSettings, and finally pass in the settings; the provider can do whatever it wants with the block, such as mapping to individual metadata. This gets the provider out of the business of interacting with the target process, which is going to be the case for egress extensibility.

Reply:
The latest change should follow what I think you're describing here - I didn't end up having the egress provider specifically say it wanted the environment block filled, so as to follow the same pattern as filling metadata (which we do regardless of egress provider), so let me know if you want me to adjust this or anything else.